src/main/java/jenkins/plugins/jfrog/BinaryInstaller.java

```diff
@@ -67,5 +100,130 @@ public String getId() {
+                    } else {
+                        log.getLogger().printf("Download '%s' version %s from: %s%n", binaryName, version, instance.getArtifactoryUrl() + cliUrlSuffix);
+                    }
+                    File file = new File(toolLocation, binaryName);
+                    Files.copy(input, file.toPath(), StandardCopyOption.REPLACE_EXISTING);
+                    if (!file.setExecutable(true)) {
+                        throw new IOException("No permission to add execution permission to binary");
+                    }
+                    createSha256File(toolLocation, client, cliUrlSuffix);
+                }
+            }
+        }
+    }
+
+    private static void createSha256File(File toolLocation, ArtifactoryClient client, String cliUrlSuffix) throws IOException {
+        // Getting cli binary's sha256 form Artifactory.
+        String artifactorySha256 = getArtifactSha256(client, cliUrlSuffix);
+        File file = new File(toolLocation, SHA256_FILE_NAME);
+        FileUtils.writeStringToFile(file, artifactorySha256, StandardCharsets.UTF_8);
+    }
+
+    /**
+     * We should skip the download if the tool's directory already contains the specific version, otherwise we should download it.
+     * A file named 'sha256' contains the specific binary sha256.
+     * If the file sha256 has not changed, we will skip the download, otherwise we will download and overwrite the existing files.
+     *
+     * @param client       - internal Artifactory Java client.
+     * @param cliUrlSuffix - path to the specific JFrog CLI version in Artifactory, will be sent to Artifactory in the HEAD request.
+     * @param toolLocation - expected location of the tool on the fileSystem.
+     */
+    private static boolean shouldDownloadTool(ArtifactoryClient client, String cliUrlSuffix, File toolLocation) throws IOException {
+        // Looking for the sha256 file in the tool directory
+        if (toolLocation == null) {
+            throw new RuntimeException("Empty tool location");
+        }
+        File[] listFiles = toolLocation.listFiles();
+        if (listFiles == null){
+            return true;
+        }
+        String sha256FileContent = "";
+        for (File file : listFiles) {
+            if (SHA256_FILE_NAME.equals(file.getName())) {
+                List<String> lines = Files.readAllLines(Paths.get(file.getPath()));
+                if (lines != null) {
+                    sha256FileContent = lines.get(0);
+                }
+            }
+        }
+        if (sha256FileContent == null || sha256FileContent.isEmpty()){
+            return true;
+        }
+        // Getting cli binary's sha256 form Artifactory.
+        String artifactorySha256 = getArtifactSha256(client, cliUrlSuffix);
+        if (artifactorySha256 != null) {
+            if (sha256FileContent.equals(artifactorySha256)) {
+                return false;
+            }
+        }
+        return true;
```

Comment:
Can be simplify to this:
```suggestion
        // Looking for the sha256 file in the tool directory
        Path path = toolLocation.toPath().resolve(SHA256_FILE_NAME);
        if (!Files.exists(path)) {
            return true;
        }
        // Getting cli binary's sha256 form Artifactory.
        String artifactorySha256 = getArtifactSha256(client, cliUrlSuffix);
        return StringUtils.equals(Files.readString(path), artifactorySha256);
```